livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java

```diff
@@ -0,0 +1,364 @@
+      if (!jsonMap.get("state").equals("idle")) {
+        while (true) {
+          LOGGER.error(String.format("sessionId:%s state is %s",
+              jsonMap.get("id"), jsonMap.get("state")));
+          Thread.sleep(1000);
+          json = executeHTTP(property.getProperty("zeppelin.livy.url") + "/sessions/" + sessionId,
+              "GET", null,
+              context.getParagraphId());
+          jsonMap = (Map<Object, Object>) gson.fromJson(json,
+              new TypeToken<Map<Object, Object>>() {
+              }.getType());
+          if (jsonMap.get("state").equals("idle")) {
+            break;
+          } else if (jsonMap.get("state").equals("error")) {
+            json = executeHTTP(property.getProperty("zeppelin.livy.url") + "/sessions/" +
+                    sessionId + "/log",
+                "GET", null,
+                context.getParagraphId());
+            jsonMap = (Map<Object, Object>) gson.fromJson(json,
+                new TypeToken<Map<Object, Object>>() {
+                }.getType());
+            String logs = StringUtils.join((ArrayList<String>) jsonMap.get("log"), '\n');
+            LOGGER.error(String.format("Cannot start  %s.\n%s", kind, logs));
+            throw new Exception(String.format("Cannot start  %s.\n%s", kind, logs));
+          }
+        }
+      }
+      return sessionId;
+    } catch (Exception e) {
+      LOGGER.error("Error getting session for user", e);
+      throw e;
+    }
+  }
+
+  protected void initializeSpark(final InterpreterContext context,
+                                 final Map<String, Integer> userSessionMap) throws Exception {
+    interpret("val sqlContext= new org.apache.spark.sql.SQLContext(sc)\n" +
+        "import sqlContext.implicits._", context, userSessionMap);
+  }
+
+  public InterpreterResult interpretInput(String stringLines,
+                                          final InterpreterContext context,
+                                          final Map<String, Integer> userSessionMap,
+                                          LivyOutputStream out) {
+    try {
+      String[] lines = stringLines.split("\n");
+      String[] linesToRun = new String[lines.length + 1];
+      for (int i = 0; i < lines.length; i++) {
+        linesToRun[i] = lines[i];
+      }
+      linesToRun[lines.length] = "print(\"\")";
+
+      out.setInterpreterOutput(context.out);
+      context.out.clear();
+      Code r = null;
+      String incomplete = "";
+
+      for (int l = 0; l < linesToRun.length; l++) {
+        String s = linesToRun[l];
+        // check if next line starts with "." (but not ".." or "./") it is treated as an invocation
+        //for spark
+        if (l + 1 < linesToRun.length) {
+          String nextLine = linesToRun[l + 1].trim();
+          if (nextLine.startsWith(".")
+              && !nextLine.startsWith("..")
+              && !nextLine.startsWith("./")) {
+            incomplete += s + "\n";
```

Comment:
you might want to check/incorporate recent changes of this code in the Spark interpreter...